### PR TITLE
Restart nlar

### PR DESCRIPTION
**Pull request by upadpup**
A basic version of restarting for NLAR(See Issue #83 ) has been implemented in src/method_nlar.jl. The function nlar() differs from the previous version in the following ways:

(1) The arguement 'mm' which specified how many ritz values should be returned by the sorter functions has been removed.

(2)Two new arguements: `max_subspace` which specifies the maximum allowable size of the subspace basis before we ought to restart and `num_restart_ritz_vecs` which specifies the number of ritz vectors that should be a part of the new basis. Hence, after each restart, the size of the basis is `num_restart_ritz_vecs + m`, where `m` is the number of eigenpairs to which we have already converged. 

(3)The variable `cbs` keeps track of the current basis size. The memory for the subspace basis matrix `V` is allocated apriori: `V::Matrix{T}= zeros(T,n,max_subspace)`. The following if-statement implements the restarting:
`if(size(Vk,2) >= max_subspace)
                cbs = m+num_restart_ritz_vecs; 

                #Construct the new basis
                Zv = view(Z,:,1:cbs);
                Zv[:,1:m] = X[:,1:m]; #Set the first m vectors of the restarted basis to be the converged eigenvectors
                Zv[:,m+1:cbs] = Vk*yv[:,1:num_restart_ritz_vecs]; #Set the rest to be the ritz vectors of the projected problem
                
                Q,_ = qr(Zv); #Thin QR-factorization
                V[:,1:cbs] = Matrix(Q);
else
                ..................
end`

### CODE REVIEW

**Comment by jarlebring** [comment]
Nice work! 

The function documentation needs some work. If you type `?nlar` you see some problems. Please make it consistent as I described in comments in #96.

**Comment by jarlebring** [comment]
At the moment you do:

1. solve proj problem
2. restart or solvlinsys+expand 

I would instead propose:
1. solve proj problem
2. solvlinsys+expand
3. restart if necessary: `if size(V...`

The projected problem solve before the restart is lost computation in the way it is currently done. Or did I miss something?

File: src/method_nlar.jl [hunk 1]
```diff
@@ -7,151 +7,84 @@ export default_eigval_sorter
...
             orthmethod::Type{T_orth} = ModifiedGramSchmidt,
             nev::Int=10,                                     #Number of eigenvalues required
-            submax_rest::Int=30,                             #Maximum dimension of subspace before restarting
             errmeasure::Function = default_errmeasure(nep),
             tol = eps(real(T))*100,
             maxit::Int = 100,
-            λ0 = zero(T),
-            v0 = randn(T,size(nep,1)),
+            λ0::Number = zero(T),
+            v0::Vector = randn(T,size(nep,1)),
```
Comment (jarlebring) [inline src/method_nlar.jl]: Please make the input consistent as we agreed in https://github.com/nep-pack/NonlinearEigenproblems.jl/wiki/Style-guidelines-and-notes i.e. λ not λ0 etc.

**Comment by jarlebring** [comment]
You are right. The if-statement is natural for me now.  :+1:

File: src/method_nlar.jl [hunk 2]
```diff
@@ -166,17 +99,19 @@ function nlar(::Type{T},
...
             # Construct and solve the small projected PEP projected problem (V^H)T(λ)Vx = 0
             set_projectmatrices!(proj_nep,Vk,Vk);
-
+            
             #Use inner_solve() to solve the smaller projected problem
             dd,vv = inner_solve(inner_solver_method,T,proj_nep,Neig=nev,σ=σ);
 
             # Sort the eigenvalues of the projected problem by measuring the distance from the eigenvalues,
             # in D and exclude all eigenvalues that lie within a unit disk of radius R from one of the
             # eigenvalues in D.
```
Comment (jarlebring) [inline src/method_nlar.jl]: This comment is too specific, since it already assumes you use a specific sorter.

File: src/method_nlar.jl [hunk 3]
```diff
@@ -194,19 +129,21 @@ function nlar(::Type{T},
 
             #Check for convergence of one of the eigenvalues
             err = errmeasure(nu,u);
-            println(k," Error:",err," Eigval :",nu)
+            if(displaylevel == 1)
+                println(k," Error:",err," Eigval :",nu)
+            end
             err_hyst[k,m+1]=err;    # Giampaolo's edit
```
Comment (jarlebring) [inline src/method_nlar.jl]: err_hyst? Shouldnt it be err_hist? as in history? The edit tag can also be removed.  Ping @meleg.

File: test/nlar.jl [hunk 4]
```diff
@@ -48,7 +48,7 @@ using Random
     c = sortperm(abs.(Dc))
     @info " 6 smallest eigenvalues found by polyeig() according to the absolute values: $(Dc[c[1:6]])"
 
-    λ,u = nlar(pepnep, tol=TOL, maxit=60, nev = 3, R=0.001,mm=1,displaylevel=1, λ0=Dc[4]+1e-2,v0=ones(size(pepnep,1)), eigval_sorter=residual_eigval_sorter,inner_solver_method=NEPSolver.IARInnerSolver,qrfact_orth=false);
+    λ,u = nlar(pepnep, tol=TOL, maxit=60, nev = 3, R=0.001,displaylevel=1, λ0=Dc[4]+1e-2,v0=ones(size(pepnep,1)), eigval_sorter=residual_eigval_sorter,inner_solver_method=NEPSolver.IARInnerSolver,qrfact_orth=false);
     @info " Smallest eigenvalues found: $λ"
```
Comment (jarlebring) [inline test/nlar.jl]: Could you add a test tagged with `@onlybench` which computes even more eiganvalues? And maybe reduce nev on the above tests.

**Comment by jarlebring** [comment]
Apart from the above minor code comments (and the merge conflict) it is ready to be merged from my perspective.

File: src/method_nlar.jl (hunk 3 above)
Comment (meleg) [reply src/method_nlar.jl]: I do not remember that I did this change. Probably there was a typo that I fixed (wrong index for example). Clearly `err_hyst` is a typo and it should be `err_hist`.

**Comment by upadpup** [comment]
TO DO: An @onlybench test to compute several eigenvalues.

**Comment by jarlebring** [comment]
nev -> neigs according to style guidelines

**Comment by meleg** [comment]
@upadpup can we delete the branch?

**Comment by upadpup** [comment]
@meleg It is deleted now.

**Comment by upadpup** [comment]
@jarlebring The arguement name is changed now.

**Pull request merged**